### PR TITLE
prov/rxm: Use fi_inject for LMT protocol

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This patch proposes to use `fi_inject()` instead `fi_send()` for transferring the following LMT protocol messages:
- Acknowledgment (1st commit)
- Request (2nd commit)

### CODE REVIEW

**Comment by shefty** [comment]
If an ack can't be sent right away, does it get queued somehow for sending later?

**Comment by a-ilango** [comment]
> If an ack can't be sent right away, does it get queued somehow for sending later?

We don't have queuing for ack in the current code.

File: prov/rxm/src/rxm_cq.c [hunk 1]
```diff
@@ -437,6 +437,35 @@ static ssize_t rxm_lmt_send_ack(struct rxm_rx_buf *rx_buf)
...
+		       "fi_inject(ack pkt) for MSG provider failed\n");
+		return ret;
+	}
+
+	RXM_LOG_STATE(FI_LOG_CQ, rx_buf->pkt, RXM_LMT_ACK_SENT, RXM_LMT_FINISH);
+	rx_buf->hdr.state = RXM_LMT_FINISH;
+
+	if (!rx_buf->ep->rxm_mr_local)
+		rxm_ep_msg_mr_closev(rx_buf->mr, RXM_IOV_LIMIT);
+	return rxm_finish_recv(rx_buf);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: Can we move the above code into a function? We have the same code in rxm_cq_handle_comp for RXM_LMT_ACK_SENT state.

**Comment by shefty** [comment]
So would the provider break if we hit this condition?  We could probably test this if verbs set the QP size to something tiny, like 4.

I wouldn't stop this PR from merging, but this is something that should be fixed if it can occur.

File: prov/rxm/src/rxm_ep.c [hunk 2]
```diff
@@ -897,8 +897,8 @@ rxm_ep_send_common(struct rxm_ep *rxm_ep, const struct iovec *iov, void **desc,
 			return -FI_EMSGSIZE;
 		}
 		ret = rxm_ep_format_tx_res(rxm_ep, rxm_conn, context, (uint8_t)count,
-					   data_len, data, flags, tag, comp_flags,
-					   &tx_buf, &tx_entry, pool);
+				   data_len, data, flags, tag, comp_flags,
+				   &tx_buf, &tx_entry, pool);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_ep.c]: nit: spacing is off

**Comment by a-ilango** [comment]
> So would the provider break if we hit this condition?

It seems we write error entry to the CQ when sending ACK fails. Perhaps we could return EAGAIN to the user so that sending the ACK could succeed in the next attempt.

**Comment by a-ilango** [comment]
But yes, we would still need a queue for storing pending ACKs.

File: prov/rxm/src/rxm_cq.c (hunk 1 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_cq.c]: Sure. Thanks
Done

File: prov/rxm/src/rxm_ep.c (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_ep.c]: Fixed

**Comment by gladkovdmitry17** [comment]
> But yes, we would still need a queue for storing pending ACKs.

Yes, it should be there.




> We could probably test this if verbs set the QP size to something tiny, like 4.

@shefty Thanks for the good idea.

**Comment by gladkovdmitry17** [comment]
The queuing of LMT protocol's messages will be implemented in the next PR.
Merging this PR so as not to stop futher work for RxM/verbs

**Pull request merged**